temporal-sdk/src/main/java/io/temporal/internal/statemachines/WorkflowStateMachines.java

```diff
@@ -171,6 +177,12 @@ public void setStartedIds(long previousStartedEventId, long workflowTaskStartedE
    * @param hasNextEvent false if this is the last event in the history.
    */
   public void handleEvent(HistoryEvent event, boolean hasNextEvent) {
+    long eventId = event.getEventId();
+    if (eventId <= lastHandledEventId) {
```

Comment:
Is it possible that this could adversely affect code paths in any backwards compatible way with existing history? I wouldn't  think so since it's fixing a cache expiration edge case, I just want to confirm from your POV.